### PR TITLE
Update to work with hmatrix 18.0.0.1 and ghc 8.2.1

### PR DESCRIPTION
**Pull request by cheater**
Would you mind looking this over? I know you don't maintain this any more but I'm specifically not sure I always used the right transpose function. I used tr' everywhere you used trans, I think that should be right. There's only one place I used tr instead of tr', and that's when we're checking that a matrix is Hermitian to use eigSH'.

Do you think the tests would catch out issues like these? Are they reasonably complete? I refactored them but I have not looked at what they actually do. Thanks.

### CODE REVIEW

File: src/HVX/Internal/DCP.hs [hunk 1]
```diff
@@ -21,77 +21,77 @@ class GetVex (v :: Vex) where
...
-  ApplyVex vf      Const  ve      me    = Affine
-  ApplyVex vf      mf     ve      Const = Affine
-  ApplyVex vf      mf     Affine  me    = vf
-  ApplyVex Affine  Nondec ve      me    = ve
-  ApplyVex Affine  Noninc ve      me    = FlipVex ve
-  ApplyVex v       Nondec v       me    = v
-  ApplyVex Convex  Noninc Concave me    = Convex
-  ApplyVex Concave Noninc Convex  me    = Concave
-  ApplyVex vf      mf     ve      me    = Nonvex
+  ApplyVex vf      'Const   ve       me     = 'Affine
```
Comment (chrisnc) [inline src/HVX/Internal/DCP.hs]: Fix the alignment of `'Const`.

**Comment by chrisnc** [comment]
I think the changes look good. I would prefer to just use `tr` (not `tr'`) everywhere, even though the library assumes only real-valued matrices. In some sense, the non-conjugate transpose is the much more specialized operation, and the conjugate transpose is more general.

**Comment by chrisnc** [comment]
As for the tests, if they build and pass, that's good enough for me. Did you try running the demo? I would expect that to produce the same results as before. I'll try out the changes sometime this week.

**Comment by chrisnc** [comment]
Thanks for contributing this, by the way!

File: src/HVX/Internal/DCP.hs (hunk 1 above)
Comment (cheater) [inline src/HVX/Internal/DCP.hs]: Am i reading this wrong or have you removed  8 lines not related to fixing the alignment?

**Comment by cheater** [comment]
The tests pass. I haven't tried the demo yet. I'll try it soon.

It seems tr' does less work than tr (correct me if I'm wrong) so if we assume real matrices we should probably be using that, right?

YW about the PR :) I'm also an uploader on hackage, is it ok if i upload hvx at some point?

When trying this make sure to use hmatrix from git. You'll have to add hmatrix/packages/base as a source to your cabal sandbox.

**Comment by cheater** [comment]
(the version of hmatrix on hackage does not compile with ghc 8.2.1)

File: src/HVX/Internal/DCP.hs (hunk 1 above)
Comment (chrisnc) [inline src/HVX/Internal/DCP.hs]: ...? I haven't changed anything, this is your PR. I was just suggesting that the `'Const` should line up with the things below it.

**Comment by chrisnc** [comment]
Re: `tr` vs `tr'`, for real matrices they should do exactly the same amount of work. By real matrices I mean that the element representation is literally only a `Double`, not a complex type, so it's not possible for there to be more work done. If/when complex numbers get involved, it will be more correct to have `tr`, so no reason not to do it now.

**Comment by chrisnc** [comment]
Feel free to upload to hackage if you like. If you do though, shouldn't you constrain the GHC version to one where the hmatrix version on hackage will work?

**Comment by cheater** [comment]
not sure how that works, i think this situation is in the hmatrix maintainer's hands

File: src/HVX/Internal/DCP.hs (hunk 1 above)
Comment (cheater) [inline src/HVX/Internal/DCP.hs]: oh, i read this wrong, sorry yeah

Comment (cheater) [inline src/HVX/Internal/DCP.hs]: fixed :) thanks for catching that!

**Comment by cheater** [comment]
I have refactored the demo code and made it executable via cabal run.

**Comment by cheater** [comment]
I haven't fixed compiler warnings on the executables. I am not sure it's worth it, i guess the main concern is for the code to be readable.

**Comment by chrisnc** [comment]
Looks good except for using `tr'` still. When writing this originally I think I assumed `trans` _was_ the conjugate transpose, because that's what you learn is the more natural way to think about Hermitian matrices, etc.

**Comment by cheater** [comment]
Just changed this

**Comment by cheater** [comment]
sorry, missed that one, changed it now

**Comment by chrisnc** [comment]
Thanks!

**Pull request merged**